File: ios/NativeCollaborationTransportRegistry.swift
import Foundation

enum NativeCollaborationTransportRegistry {
    typealias EventEmitter = ([String: Any]) -> Void

    private static let queue = DispatchQueue(
        label: "com.apollohg.native-editor.collaboration-registry"
    )
    private static var transports: [UInt64: NativeCollaborationTransport] = [:]
    private static var transportOwners: [UInt64: UUID] = [:]
    private static var transportTokens: [UInt64: UUID] = [:]
    private static var eventSequences: [UInt64: UInt64] = [:]
    private static var eventProcessors: [UInt64: CollaborationEventProcessor] = [:]
    private static var remoteRebases: [UInt64: RemoteRebaseState] = [:]
    private static var eventEmitter: EventEmitter?
    private static var eventEmitterOwner: UUID?

    private struct RemoteRebaseState {
        let token: UUID
        var dirty: Bool
    }

    private final class CollaborationEventProcessor {
        let queue: DispatchQueue

        init(editorId: UInt64) {
            queue = DispatchQueue(
                label: "com.apollohg.native-editor.collaboration-events.\(editorId)"
            )
        }
    }

    static func setEventEmitter(owner: UUID, _ emitter: EventEmitter?) {
        queue.sync {
            if let emitter {
                eventEmitter = emitter
                eventEmitterOwner = owner
            } else if eventEmitterOwner == owner {
                eventEmitter = nil
                eventEmitterOwner = nil
            }
        }
    }

    static func configure(owner: UUID, editorId: UInt64, configJSON: String?) -> FfiError? {
        guard editorId > 0 else {
            return contractError("invalid editorId")
        }
        let parsed: NativeCollaborationTransportConfig?
        if let configJSON {
            switch parseConfig(configJSON) {
            case .success(let config):
                parsed = config
            case .failure(let error):
                return error
            }
        } else {
            parsed = nil
        }

        return queue.sync {
            if parsed == nil {
                let existing = transports.removeValue(forKey: editorId)
                transportOwners.removeValue(forKey: editorId)
                transportTokens.removeValue(forKey: editorId)
                eventProcessors.removeValue(forKey: editorId)
                remoteRebases.removeValue(forKey: editorId)
                existing?.destroy()
                return nil
            }

            let transport: NativeCollaborationTransport
            let created: Bool
            if let existing = transports[editorId] {
                transport = existing
                created = false
            } else {
                let token = UUID()
                transport = NativeCollaborationTransport(
                    editorId: String(editorId),
                    eventSink: { event in
                        enqueue(event: event, editorId: editorId, token: token)
                    }
                )
                transports[editorId] = transport
                transportOwners[editorId] = owner
                transportTokens[editorId] = token
                eventProcessors[editorId] = CollaborationEventProcessor(editorId: editorId)
                created = true
            }
            transportOwners[editorId] = owner
            let error = transport.configure(parsed)
            if error != nil, created, transports[editorId] === transport {
                // A rejected Rust lifecycle transition must not leave a new
                // unusable owner registered. Existing owners remain intact.
                transports.removeValue(forKey: editorId)
                transportOwners.removeValue(forKey: editorId)
                transportTokens.removeValue(forKey: editorId)
                eventProcessors.removeValue(forKey: editorId)
                transport.destroy()
            }
            return error
        }
    }

    static func notifyOutboundAvailable(editorId: UInt64, reason: CollaborationWakeReason) {
        queue.async {
            transports[editorId]?.notifyOutboundAvailable(reason: reason)
        }
    }

    static func resolveProtocolAdapter(
        editorId: UInt64,
        attemptId: String,
        eventId: String,
        responseJSON: String
    ) -> FfiError? {
        guard editorId > 0,
              !attemptId.isEmpty,
              canonicalProtocolEventId(eventId) != nil,
              let response = parseProtocolAdapterResponse(responseJSON)
        else {
            return contractError("invalid collaboration protocol adapter response")
        }
        return queue.sync {
            guard let transport = transports[editorId] else {
                // A response racing teardown belongs to a retired attempt.
                return nil
            }
            return transport.resolveProtocolAdapter(
                attemptId: attemptId,
                eventId: eventId,
                response: response
            )
        }
    }

    static func destroy(editorId: UInt64) {
        queue.sync {
            transports.removeValue(forKey: editorId)?.destroy()
            transportOwners.removeValue(forKey: editorId)
            transportTokens.removeValue(forKey: editorId)
            eventSequences.removeValue(forKey: editorId)
            eventProcessors.removeValue(forKey: editorId)
            remoteRebases.removeValue(forKey: editorId)
        }
    }

    static func enterBackground() {
        queue.sync {
            transports.values.forEach { $0.enterBackground() }
        }
    }

    static func enterForeground() {
        queue.sync {
            transports.values.forEach { $0.enterForeground() }
        }
    }

    static func destroyAll(owner: UUID) {
        queue.sync {
            let ownedEditorIds = transportOwners.compactMap { editorId, registeredOwner in
                registeredOwner == owner ? editorId : nil
            }
            let owned = ownedEditorIds.compactMap { transports.removeValue(forKey: $0) }
            for editorId in ownedEditorIds {
                transportOwners.removeValue(forKey: editorId)
                transportTokens.removeValue(forKey: editorId)
                eventSequences.removeValue(forKey: editorId)
                eventProcessors.removeValue(forKey: editorId)
                remoteRebases.removeValue(forKey: editorId)
            }
            owned.forEach { $0.destroy() }
            if eventEmitterOwner == owner {
                eventEmitter = nil
                eventEmitterOwner = nil
            }
        }
    }

#if DEBUG
    static var eventEmitterOwnerForTesting: UUID? {
        queue.sync { eventEmitterOwner }
    }

    static func emitForTesting(_ payload: [String: Any]) {
        queue.sync { eventEmitter?(payload) }
    }
#endif

    private static func enqueue(
        event: NativeCollaborationTransportEvent,
        editorId: UInt64,
        token: UUID
    ) {
        queue.async {
            guard transports[editorId] != nil,
                  transportTokens[editorId] == token,
                  let processor = eventProcessors[editorId]
            else { return }
            let next = (eventSequences[editorId] ?? 0).addingReportingOverflow(1)
            guard !next.overflow, next.partialValue > 0 else {
                return
            }
            eventSequences[editorId] = next.partialValue
            let sequence = next.partialValue
            processor.queue.async {
                var payload: [String: Any] = [
                    "editorId": String(editorId),
                    "eventSequence": String(sequence)
                ]
                switch event {
                case let .directive(directive, generation, wakeReason):
                    guard let state = state(editorId: editorId) else { return }
                    payload["kind"] = "state"
                    payload["generation"] = generation ?? NSNull()
                    payload["state"] = state
                    payload["peers"] = peers(editorId: editorId)
                    payload["diagnostics"] = [
                        "wakeReason": wakeReason.rawValue,
                        "transportState": directive.transportState,
                        "nextDeadlineMillis": directive.nextDeadlineMillis ?? NSNull(),
                        "remoteCommitApplied": directive.remoteCommitApplied,
                        "peersChanged": directive.peersChanged,
                        "renewedLocal": directive.renewedLocal,
                        "expiredPeerCount": directive.expiredPeers.count
                    ]
                case let .error(error, generation):
                    payload["kind"] = "error"
                    payload["generation"] = generation ?? NSNull()
                    payload["error"] = errorDictionary(error)
                case let .protocolAdapter(adapterEvent):
                    payload["kind"] = "protocolAdapter"
                    payload["generation"] = adapterEvent.generation
                    payload["attemptId"] = adapterEvent.attemptId
                    payload["eventId"] = adapterEvent.eventId
                    payload["negotiatedProtocol"] =
                        adapterEvent.negotiatedProtocol ?? NSNull()
                    switch adapterEvent.phase {
                    case .open:
                        payload["phase"] = "open"
                    case .message(.text(let text)):
                        payload["phase"] = "message"
                        payload["frame"] = ["type": "text", "data": text]
                    case .message(.binary(let data)):
                        payload["phase"] = "message"
                        payload["frame"] = [
                            "type": "binary",
                            "data": data.base64EncodedString()
                        ]
                    }
                }
                queue.async {
                    guard transports[editorId] != nil,
                          transportTokens[editorId] == token,
                          eventProcessors[editorId] === processor
                    else { return }
                    if case let .directive(directive, _, _) = event,
                       directive.remoteCommitApplied {
                        scheduleRemoteRebase(editorId: editorId, token: token)
                    }
                    eventEmitter?(payload)
                }
            }
        }
    }

    private static func scheduleRemoteRebase(editorId: UInt64, token: UUID) {
        if var pending = remoteRebases[editorId], pending.token == token {
            pending.dirty = true
            remoteRebases[editorId] = pending
            return
        }
        remoteRebases[editorId] = RemoteRebaseState(token: token, dirty: false)
        dispatchRemoteRebase(editorId: editorId, token: token)
    }

    private static func dispatchRemoteRebase(editorId: UInt64, token: UUID) {
        DispatchQueue.main.async {
            NativeEditorViewRegistry.shared.applyRemoteCommitRefresh(editorId: editorId)
            queue.async {
                guard var pending = remoteRebases[editorId], pending.token == token else { return }
                if pending.dirty {
                    pending.dirty = false
                    remoteRebases[editorId] = pending
                    dispatchRemoteRebase(editorId: editorId, token: token)
                } else {
                    remoteRebases.removeValue(forKey: editorId)
                }
            }
        }
    }

    private static func state(editorId: UInt64) -> [String: Any]? {
        let handle = String(editorId)
        let result = performEditorV2JsonOperation(editorId: handle) {
            editorV2GetState(editorId: handle)
        }
        guard result.error == nil,
              let value = result.value,
              let data = value.data(using: .utf8),
              let decoded = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return nil
        }
        return decoded
    }

    private static func peers(editorId: UInt64) -> Any {
        let handle = String(editorId)
        let result = performEditorV2JsonOperation(editorId: handle) {
            editorV2CollaborationPeers(editorId: handle)
        }
        guard result.error == nil,
              let value = result.value,
              let data = value.data(using: .utf8),
              let decoded = try? JSONSerialization.jsonObject(with: data)
        else {
            return []
        }
        if let object = decoded as? [String: Any],
           let peers = object["peers"] {
            return peers
        }
        return []
    }

    private static func parseConfig(
        _ json: String
    ) -> Result<NativeCollaborationTransportConfig, FfiError> {
        guard let data = json.data(using: .utf8),
              data.count <= 32_768,
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              Set(object.keys).isSubset(of: Set(["url", "connect", "protocolAdapter"])),
              object.keys.contains("url"),
              object.keys.contains("connect"),
              let rawURL = object["url"] as? String,
              let connectNumber = object["connect"] as? NSNumber,
              CFGetTypeID(connectNumber) == CFBooleanGetTypeID(),
              let url = URL(string: rawURL)
        else {
            return .failure(contractError("invalid collaboration transport configuration"))
        }
        let protocolAdapter: NativeCollaborationProtocolAdapterConfig?
        if let rawAdapter = object["protocolAdapter"] {
            guard let adapter = parseProtocolAdapterConfig(rawAdapter) else {
                return .failure(contractError("invalid collaboration protocol adapter"))
            }
            protocolAdapter = adapter
        } else {
            protocolAdapter = nil
        }
        do {
            return .success(try NativeCollaborationTransportConfig(
                url: url,
                connect: connectNumber.boolValue,
                protocolAdapter: protocolAdapter
            ))
        } catch {
            return .failure(contractError("invalid collaboration WebSocket URL"))
        }
    }

    private static func parseProtocolAdapterConfig(
        _ raw: Any
    ) -> NativeCollaborationProtocolAdapterConfig? {
        guard let object = raw as? [String: Any],
              Set(object.keys).isSubset(of: Set([
                "protocols",
                "timeoutMillis",
                "terminalCloseCodes"
              ])),
              let protocols = object["protocols"] as? [String],
              !protocols.isEmpty,
              protocols.count <= 16,
              Set(protocols).count == protocols.count,
              protocols.allSatisfy(validWebSocketProtocol)
        else {
            return nil
        }
        let timeoutMillis: UInt64
        if let rawTimeout = object["timeoutMillis"] {
            guard let number = rawTimeout as? NSNumber,
                  CFGetTypeID(number) != CFBooleanGetTypeID(),
                  number.doubleValue.rounded(.towardZero) == number.doubleValue,
                  number.doubleValue >= 1,
                  number.doubleValue <= 60_000
            else {
                return nil
            }
            timeoutMillis = number.uint64Value
        } else {
            timeoutMillis = NativeCollaborationProtocolAdapterConfig.defaultTimeoutMillis
        }
        let terminalCloseCodes: Set<UInt32>
        if let rawCodes = object["terminalCloseCodes"] {
            guard let values = rawCodes as? [Any] else { return nil }
            var codes = Set<UInt32>()
            for rawCode in values {
                guard let number = rawCode as? NSNumber,
                      CFGetTypeID(number) != CFBooleanGetTypeID(),
                      number.doubleValue.rounded(.towardZero) == number.doubleValue,
                      number.doubleValue >= 1_000,
                      number.doubleValue <= 4_999
                else {
                    return nil
                }
                guard codes.insert(number.uint32Value).inserted else { return nil }
            }
            terminalCloseCodes = codes
        } else {
            terminalCloseCodes = []
        }
        return NativeCollaborationProtocolAdapterConfig(
            protocols: protocols,
            timeoutMillis: timeoutMillis,
            terminalCloseCodes: terminalCloseCodes
        )
    }

    private static func validWebSocketProtocol(_ value: String) -> Bool {
        guard let bytes = value.data(using: .utf8),
              !bytes.isEmpty,
              bytes.count <= 128
        else {
            return false
        }
        let allowed = CharacterSet(
            charactersIn: "!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
        )
        return value.unicodeScalars.allSatisfy(allowed.contains)
    }

    private static func parseProtocolAdapterResponse(
        _ json: String
    ) -> NativeCollaborationProtocolAdapterResponse? {
        guard let data = json.data(using: .utf8),
              data.count <= 1_500_000,
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              Set(object.keys).isSubset(of: Set(["action", "frames"])),
              let rawAction = object["action"] as? String
        else {
            return nil
        }
        let action: NativeCollaborationProtocolAdapterAction
        switch rawAction {
        case "continue": action = .continue
        case "ready": action = .ready
        case "reject": action = .reject
        default: return nil
        }
        let rawFrames = object["frames"] as? [Any] ?? []
        guard rawFrames.count <= 16 else { return nil }
        var frames: [NativeCollaborationProtocolFrame] = []
        frames.reserveCapacity(rawFrames.count)
        for rawFrame in rawFrames {
            guard let frame = rawFrame as? [String: Any],
                  Set(frame.keys) == Set(["type", "data"]),
                  let type = frame["type"] as? String,
                  let value = frame["data"] as? String
            else {
                return nil
            }
            switch type {
            case "text":
                guard (value.data(using: .utf8)?.count ?? Int.max) <= 64 * 1_024 else {
                    return nil
                }
                frames.append(.text(value))
            case "binary":
                guard let decoded = Data(base64Encoded: value),
                      decoded.count <= 64 * 1_024
                else {
                    return nil
                }
                frames.append(.binary(decoded))
            default:
                return nil
            }
        }
        return NativeCollaborationProtocolAdapterResponse(action: action, frames: frames)
    }

    private static func canonicalProtocolEventId(_ raw: String) -> UInt64? {
        guard !raw.isEmpty,
              raw.allSatisfy({ $0 >= "0" && $0 <= "9" }),
              raw == "0" || raw.first != "0"
        else {
            return nil
        }
        return UInt64(raw)
    }

    private static func errorDictionary(_ error: FfiError) -> [String: Any] {
        var value: [String: Any] = [
            "domain": error.domain,
            "code": error.code,
            "message": error.message
        ]
        if let requestId = error.requestId { value["requestId"] = requestId }
        if let operationIndex = error.operationIndex { value["operationIndex"] = operationIndex }
        if let limit = error.limit { value["limit"] = limit }
        if let actual = error.actual { value["actual"] = actual }
        if let detailsJson = error.detailsJson { value["detailsJson"] = detailsJson }
        return value
    }

    private static func contractError(_ message: String) -> FfiError {
        FfiError(
            domain: "boundary",
            code: "FFI_RESULT_INVALID",
            message: message,
            requestId: nil,
            operationIndex: nil,
            limit: nil,
            actual: nil,
            detailsJson: nil
        )
    }
}

File: android/src/main/java/com/apollohg/editor/NativeCollaborationTransportRegistry.kt
package com.apollohg.editor

import android.util.Base64
import java.util.concurrent.CompletableFuture
import org.json.JSONArray
import org.json.JSONObject

internal object NativeCollaborationTransportRegistry {
    private val lock = Any()
    private val transports = mutableMapOf<String, AndroidCollaborationTransport>()
    private val transportTokens = mutableMapOf<String, Any>()
    private val retirements = mutableMapOf<String, CompletableFuture<Unit>>()
    private val eventSequences = mutableMapOf<String, ULong>()
    private val configureLocks = Array(64) { Any() }
    private var eventEmitter: ((Map<String, Any?>) -> Unit)? = null
    private var hostState = AndroidCollaborationTransport.HostState.FOREGROUND
    private var runtimeActive = true
    private var runtimeToken = Any()

    @Volatile
    internal var transportFactoryForTesting: (
        (
            String,
            (AndroidCollaborationTransportEvent) -> Unit
        ) -> AndroidCollaborationTransport
    )? = null

    fun setEventEmitter(ownerToken: Any, emitter: ((Map<String, Any?>) -> Unit)?) {
        synchronized(lock) {
            if (!runtimeActive || runtimeToken !== ownerToken) return
            eventEmitter = emitter
        }
    }

    fun activateRuntime(): Any = synchronized(lock) {
        Any().also {
            runtimeActive = true
            runtimeToken = it
        }
    }

    fun configure(ownerToken: Any?, editorId: String, configJson: String?): EditorV2Error? {
        synchronized(lock) {
            if (!runtimeActive || runtimeToken !== ownerToken) return runtimeDestroyedError()
        }
        val canonical = canonicalV2U64(editorId)?.takeIf { it != "0" }
            ?: return contractError("invalid editorId")
        val config = if (configJson == null) {
            null
        } else {
            parseConfig(configJson) ?: return contractError(
                "invalid collaboration transport configuration"
            )
        }

        return synchronized(configureLock(canonical)) {
            val operationToken = synchronized(lock) {
                if (!runtimeActive || runtimeToken !== ownerToken) return@synchronized null
                ownerToken
            } ?: return@synchronized runtimeDestroyedError()
            if (config == null) {
                synchronized(lock) {
                    transportTokens.remove(canonical)
                    transports.remove(canonical)?.let { retireLocked(canonical, it) }
                }
                return@synchronized null
            }
            awaitRetirement(canonical)
            val owned = synchronized(lock) {
                if (!runtimeActive || runtimeToken !== operationToken) return@synchronized null
                val created = !transports.containsKey(canonical)
                val transport = transports.getOrPut(canonical) {
                    val token = Any()
                    transportTokens[canonical] = token
                    val sink = { event: AndroidCollaborationTransportEvent ->
                        enqueueEvent(canonical, token, event)
                    }
                    (
                        transportFactoryForTesting?.invoke(canonical, sink)
                            ?: AndroidCollaborationTransport(editorId = canonical, eventSink = sink)
                        ).also {
                        it.requestHostState(hostState)
                    }
                }
                transport to created
            } ?: return@synchronized runtimeDestroyedError()
            val (transport, created) = owned
            val error = transport.configure(config)
            val stillCurrent = synchronized(lock) {
                runtimeActive && runtimeToken === operationToken &&
                    transports[canonical] === transport
            }
            if (!stillCurrent) return@synchronized runtimeDestroyedError()
            if (error != null && created) {
                synchronized(lock) {
                    if (transports[canonical] === transport) {
                        transportTokens.remove(canonical)
                        transports.remove(canonical)?.let { retireLocked(canonical, it) }
                    }
                }
            }
            error
        }
    }

    fun notifyOutboundAvailable(editorId: String, reason: CollaborationWakeReason) {
        val canonical = canonicalV2U64(editorId) ?: return
        synchronized(lock) {
            transports[canonical]
        }?.notifyOutboundAvailable(reason)
    }

    fun resolveProtocolAdapter(
        ownerToken: Any?,
        editorId: String,
        attemptId: String,
        eventId: String,
        responseJson: String
    ): EditorV2Error? {
        synchronized(lock) {
            if (!runtimeActive || runtimeToken !== ownerToken) return runtimeDestroyedError()
        }
        val canonical = canonicalV2U64(editorId)?.takeIf { it != "0" }
            ?: return contractError("invalid editorId")
        val response = parseProtocolAdapterResponse(responseJson)
            ?: return contractError("invalid collaboration protocol adapter response")
        if (attemptId.isEmpty() || canonicalV2U64(eventId) == null) {
            return contractError("invalid collaboration protocol adapter response")
        }
        val transport = synchronized(lock) {
            if (!runtimeActive || runtimeToken !== ownerToken) return runtimeDestroyedError()
            transports[canonical]
        } ?: return null
        return transport.resolveProtocolAdapter(attemptId, eventId, response)
    }

    fun destroy(editorId: String) {
        val canonical = canonicalV2U64(editorId) ?: return
        synchronized(configureLock(canonical)) {
            synchronized(lock) {
                transportTokens.remove(canonical)
                transports.remove(canonical)?.let { retireLocked(canonical, it) }
                eventSequences.remove(canonical)
            }
        }
    }

    fun enterBackground(ownerToken: Any) = requestHostState(
        ownerToken,
        AndroidCollaborationTransport.HostState.BACKGROUND
    )

    fun attachHost(ownerToken: Any) = requestHostState(
        ownerToken,
        AndroidCollaborationTransport.HostState.FOREGROUND
    )

    fun detachHost(ownerToken: Any) = requestHostState(
        ownerToken,
        AndroidCollaborationTransport.HostState.DETACHED
    )

    private fun requestHostState(
        ownerToken: Any,
        requestedState: AndroidCollaborationTransport.HostState
    ) {
        val owned = synchronized(lock) {
            if (!runtimeActive || runtimeToken !== ownerToken) return
            hostState = requestedState
            transports.values.toList()
        }
        owned.forEach {
            it.requestHostState(requestedState)
        }
    }

    fun destroyRuntime(ownerToken: Any) {
        synchronized(lock) {
            if (!runtimeActive || runtimeToken !== ownerToken) return
            runtimeActive = false
            runtimeToken = Any()
            transports.forEach { (editorId, transport) -> retireLocked(editorId, transport) }
            transports.clear()
            transportTokens.clear()
            eventSequences.clear()
            eventEmitter = null
            hostState = AndroidCollaborationTransport.HostState.DETACHED
        }
    }

    private fun configureLock(editorId: String): Any =
        configureLocks[(editorId.hashCode() and Int.MAX_VALUE) % configureLocks.size]

    private fun runtimeDestroyedError() =
        EditorV2Error("lifecycle", "ENGINE_DESTROYED", "collaboration runtime is destroyed")

    private fun retireLocked(editorId: String, transport: AndroidCollaborationTransport) {
        val completion = transport.destroyAsync()
        retirements[editorId] = completion
        completion.whenComplete { _, _ ->
            synchronized(lock) {
                if (retirements[editorId] === completion) retirements.remove(editorId)
            }
        }
    }

    private fun awaitRetirement(editorId: String) {
        while (true) {
            val completion = synchronized(lock) { retirements[editorId] } ?: return
            runCatching { completion.get() }
        }
    }

    internal fun containsForTesting(editorId: String): Boolean {
        val canonical = canonicalV2U64(editorId) ?: return false
        return synchronized(lock) { canonical in transports }
    }

    internal fun identityForTesting(editorId: String): Any? {
        val canonical = canonicalV2U64(editorId) ?: return null
        return synchronized(lock) { transports[canonical] }
    }

    internal fun hasEventEmitterForTesting(): Boolean = synchronized(lock) {
        eventEmitter != null
    }

    internal fun awaitIdleForTesting(editorId: String) {
        val canonical = canonicalV2U64(editorId) ?: return
        synchronized(lock) { transports[canonical] }?.awaitIdleForTesting()
    }

    internal fun emitErrorForTesting(editorId: String, error: EditorV2Error) {
        val canonical = canonicalV2U64(editorId) ?: return
        val token = synchronized(lock) { transportTokens[canonical] } ?: return
        enqueueEvent(canonical, token, AndroidCollaborationTransportEvent.Error(error, null))
    }

    private fun enqueueEvent(
        editorId: String,
        token: Any,
        event: AndroidCollaborationTransportEvent
    ) {
        val sequence = synchronized(lock) {
            if (!transports.containsKey(editorId) || transportTokens[editorId] !== token) return
            val current = eventSequences[editorId] ?: 0uL
            if (current == ULong.MAX_VALUE) return
            val next = current + 1uL
            eventSequences[editorId] = next
            next
        }
        val payload = mutableMapOf<String, Any?>(
            "editorId" to editorId,
            "eventSequence" to sequence.toString()
        )
        when (event) {
            is AndroidCollaborationTransportEvent.Directive -> {
                val state = state(editorId) ?: return
                payload["kind"] = "state"
                payload["generation"] = event.generation
                payload["state"] = state
                payload["peers"] = peers(editorId)
                payload["diagnostics"] = mapOf(
                    "wakeReason" to event.wakeReason.wireValue,
                    "transportState" to event.directive.transportState,
                    "nextDeadlineMillis" to event.directive.nextDeadlineMillis,
                    "remoteCommitApplied" to event.directive.remoteCommitApplied,
                    "peersChanged" to event.directive.peersChanged,
                    "renewedLocal" to event.directive.renewedLocal,
                    "expiredPeerCount" to event.directive.expiredPeers.size
                )
            }

            is AndroidCollaborationTransportEvent.Error -> {
                payload["kind"] = "error"
                payload["generation"] = event.generation
                payload["error"] = event.error.toJSMap()
            }

            is AndroidCollaborationTransportEvent.ProtocolAdapter -> {
                val adapterEvent = event.event
                payload["kind"] = "protocolAdapter"
                payload["generation"] = adapterEvent.generation
                payload["attemptId"] = adapterEvent.attemptId
                payload["eventId"] = adapterEvent.eventId
                payload["negotiatedProtocol"] = adapterEvent.negotiatedProtocol
                when (val phase = adapterEvent.phase) {
                    NativeCollaborationProtocolAdapterPhase.Open -> {
                        payload["phase"] = "open"
                    }

                    is NativeCollaborationProtocolAdapterPhase.Message -> {
                        payload["phase"] = "message"
                        payload["frame"] = when (val frame = phase.frame) {
                            is NativeCollaborationProtocolFrame.Text ->
                                mapOf("type" to "text", "data" to frame.data)

                            is NativeCollaborationProtocolFrame.Binary ->
                                mapOf(
                                    "type" to "binary",
                                    "data" to Base64.encodeToString(frame.data, Base64.NO_WRAP)
                                )
                        }
                    }
                }
            }
        }
        val emitter = synchronized(lock) {
            if (!transports.containsKey(editorId) || transportTokens[editorId] !== token) return
            eventEmitter
        }
        if (event is AndroidCollaborationTransportEvent.Directive &&
            event.directive.remoteCommitApplied
        ) {
            NativeEditorViewRegistry.rebaseAfterRemoteCommit(editorId)
        }
        emitter?.invoke(payload)
    }

    private fun state(editorId: String): Map<String, Any?>? {
        val result = UniffiEditorV2Backend.getState(editorId)
        if (result !is EditorV2CallResult.Ok) return null
        return runCatching { JSONObject(result.value).toJsMap() }.getOrNull()
    }

    private fun peers(editorId: String): Any {
        val result = UniffiEditorV2Backend.collaborationPeers(editorId)
        if (result !is EditorV2CallResult.Ok) return emptyList<Any>()
        return runCatching {
            val objectValue = JSONObject(result.value)
            jsonValueToJs(objectValue.getJSONArray("peers")) ?: emptyList<Any>()
        }.getOrDefault(emptyList<Any>())
    }

    private fun parseConfig(json: String): NativeCollaborationTransportConfig? {
        if (json.toByteArray(Charsets.UTF_8).size > 32_768) return null
        val value = runCatching { JSONObject(json) }.getOrNull() ?: return null
        val keys = value.keys().asSequence().toSet()
        if (
            !keys.containsAll(setOf("url", "connect")) ||
            !setOf("url", "connect", "protocolAdapter").containsAll(keys)
        ) {
            return null
        }
        val url = value.opt("url") as? String ?: return null
        val connect = value.opt("connect") as? Boolean ?: return null
        val protocolAdapter = when (val rawAdapter = value.opt("protocolAdapter")) {
            null -> null
            is JSONObject -> parseProtocolAdapterConfig(rawAdapter) ?: return null
            else -> return null
        }
        return NativeCollaborationTransportConfig.parse(url, connect, protocolAdapter)
    }

    private fun parseProtocolAdapterConfig(
        value: JSONObject
    ): NativeCollaborationProtocolAdapterConfig? {
        val keys = value.keys().asSequence().toSet()
        if (
            !keys.contains("protocols") ||
            !setOf("protocols", "timeoutMillis", "terminalCloseCodes").containsAll(keys)
        ) {
            return null
        }
        val rawProtocols = value.optJSONArray("protocols") ?: return null
        if (rawProtocols.length() !in 1..16) return null
        val protocols = List(rawProtocols.length()) { index ->
            rawProtocols.opt(index) as? String ?: return null
        }
        if (protocols.toSet().size != protocols.size ||
            protocols.any { !validWebSocketProtocol(it) }
        ) {
            return null
        }
        val timeoutMillis = when (val rawTimeout = value.opt("timeoutMillis")) {
            null -> 10_000L

            is Number -> {
                val asDouble = rawTimeout.toDouble()
                if (
                    !asDouble.isFinite() ||
                    asDouble % 1.0 != 0.0 ||
                    asDouble < 1.0 ||
                    asDouble > 60_000.0
                ) {
                    return null
                }
                asDouble.toLong()
            }

            else -> return null
        }
        val terminalCloseCodes = when (val rawCodes = value.optJSONArray("terminalCloseCodes")) {
            null -> emptySet()

            else -> buildSet {
                for (index in 0 until rawCodes.length()) {
                    val rawCode = rawCodes.opt(index) as? Number ?: return null
                    val asDouble = rawCode.toDouble()
                    if (
                        !asDouble.isFinite() ||
                        asDouble % 1.0 != 0.0 ||
                        asDouble < 1_000.0 ||
                        asDouble > 4_999.0 ||
                        !add(asDouble.toInt())
                    ) {
                        return null
                    }
                }
            }
        }
        return NativeCollaborationProtocolAdapterConfig(
            protocols = protocols,
            timeoutMillis = timeoutMillis,
            terminalCloseCodes = terminalCloseCodes
        )
    }

    private fun validWebSocketProtocol(value: String): Boolean {
        if (value.toByteArray(Charsets.UTF_8).size !in 1..128) return false
        val allowed =
            "!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
                .toSet()
        return value.all { it in allowed }
    }

    private fun parseProtocolAdapterResponse(
        json: String
    ): NativeCollaborationProtocolAdapterResponse? {
        if (json.toByteArray(Charsets.UTF_8).size > 1_500_000) return null
        val value = runCatching { JSONObject(json) }.getOrNull() ?: return null
        val keys = value.keys().asSequence().toSet()
        if (!keys.contains("action") || !setOf("action", "frames").containsAll(keys)) return null
        val action = when (value.opt("action") as? String) {
            "continue" -> NativeCollaborationProtocolAdapterAction.CONTINUE
            "ready" -> NativeCollaborationProtocolAdapterAction.READY
            "reject" -> NativeCollaborationProtocolAdapterAction.REJECT
            else -> return null
        }
        val rawFrames = value.optJSONArray("frames")
        if (rawFrames != null && rawFrames.length() > 16) return null
        val frames = buildList {
            if (rawFrames != null) {
                for (index in 0 until rawFrames.length()) {
                    val frame = rawFrames.optJSONObject(index) ?: return null
                    if (frame.keys().asSequence().toSet() != setOf("type", "data")) return null
                    val data = frame.opt("data") as? String ?: return null
                    when (frame.opt("type") as? String) {
                        "text" -> {
                            if (
                                data.toByteArray(Charsets.UTF_8).size >
                                NativeCollaborationProtocolAdapterConfig.MAXIMUM_FRAME_BYTES
                            ) {
                                return null
                            }
                            add(NativeCollaborationProtocolFrame.Text(data))
                        }

                        "binary" -> {
                            val decoded = runCatching {
                                Base64.decode(data, Base64.NO_WRAP)
                            }.getOrNull() ?: return null
                            if (
                                decoded.size >
                                NativeCollaborationProtocolAdapterConfig.MAXIMUM_FRAME_BYTES
                            ) {
                                return null
                            }
                            add(NativeCollaborationProtocolFrame.Binary(decoded))
                        }

                        else -> return null
                    }
                }
            }
        }
        return NativeCollaborationProtocolAdapterResponse(action, frames)
    }

    private fun contractError(message: String) =
        EditorV2Error("boundary", "FFI_RESULT_INVALID", message)
}

private fun JSONObject.toJsMap(): Map<String, Any?> =
    keys().asSequence().associateWith { key -> jsonValueToJs(get(key)) }

internal fun jsonValueToJs(value: Any?): Any? = when (value) {
    null, JSONObject.NULL -> null
    is JSONObject -> value.toJsMap()
    is JSONArray -> List(value.length()) { index -> jsonValueToJs(value.get(index)) }
    is String, is Boolean, is Number -> value
    else -> throw IllegalArgumentException("unsupported JSON value")
}
